### ISSUE TITLE
Global search

### ISSUE THREAD

**Issue by 1337sup3rh4x0r** [owncloud/mail#1044]
It would be great to have the ability to use global search, including all subfolders.

Maybe even an integration of full text search into the general search function of owncloud. Or am I dreaming now? =)

**Comment by jancborchardt** [comment]
Yeah, that would really be cool. Priority can be on the current folder, and then a request can be sent off in the background to load more results from other folders.

> Maybe even an integration of full text search into the general search function of owncloud.

What do you mean by that? This is already the case. :)

**Comment by 1337sup3rh4x0r** [comment]
> What do you mean by that? This is already the case. :)

For example when I use the search box from within the file app, I not only get file results, but also events and contacts and so on. Would be great if also emails containing the word would be listed there.

**Comment by DeepDiver1975** [comment]
This wil be really challencing. We would need to fire one search against each folder/subfolder. A performance nightmare.

No IMAP client does that afaik.

**Comment by pravussum** [comment]
I think searching via IMAP server will never be fast enough. Even my normal search (using Dovecot on a vServer) is almost unusable, since it starts to refresh the index on the first search.

What about a background task continously building up a lucence index with corresponding message UIDs? 
Did this in my own java based mail client and it works quite well and really fast. Most challenging part is compatibility with all kind of IMAP servers since not all allow search by message UID afaik. 

If this problem is solved this would allow for other features like message tags etc.

**Comment by PoPoutdoor** [comment]
Global search with desktop client only search for downloaded mails. For Mail app, local mail storage is needed.

Can we use File app api to construct storage for downloaded Mail?

**Comment by PoPoutdoor** [comment]
When we had local mail storage, thread view can be done in a easier way.

**Comment by pravussum** [comment]
I am not talking about storing the whole mail locally, only metadata (tags, search index, etc.). 
Storing the whole mail would be a waste of space especially when mailserver and owncloud are running on the same machine.

**Comment by PoPoutdoor** [comment]
@pravussum IMHO, your use case seems out of scope for Mail app.

**Comment by jancborchardt** [comment]
Yeah. But global search (that is, lazily showing results of other folders, especially important ones such as Inbox and Sent) is definitely in scope, although for the future.

**Comment by pravussum** [comment]
@PoPoutdoor 
Why is that out of scope? Slow search is the main reason why all the webbased mail clients out there suck. Even Horde and roundcube rely on the IMAP server for their searches und are just sooo slow. 
Gmail does is right, but has it's own mail format. Thunderbird does it right, but has a local copy.
My suggestions just brings these two things together: the flexibility of using existing IMAP servers without keeping a local copy but still having a fast search via a local index.
Plus the bonus of creating the foundation for some cool future features (tags, reminders, etc. etc.) to make a difference. 
Or why else should you go through the pain of recreating another email app?

**Comment by PoPoutdoor** [comment]
@pravussum Your idea is better done at mail server side by providing client some search API. For Mail app will never be both server & client, that's why I said "out of scope"

**Comment by faucon** [comment]
I just found this discussion and wanted to comment on this:
Yes, Roundcube relies on the mail server, but at least it provides the ability to search all folders.
And yes, on most servers this will be slow, but this problem can be handled on the mail server side:

I set up a full text index with dovecot (http://wiki.dovecot.org/Plugins/FTS) and then global mail search in roundcube works increadibly fast.

So why not delegate the (fast) searching to the mail sever and still implement the feature of searching all folders?
(As far as I know there is (not yet) an IMAP standard to search multiple folders at once, but you just have to issue IMAP commands per folder).

For my this feature is one of the reasons to still use rounducbe instead of owncloud mail.

**Comment by ChristophWurst** [comment]
@faucon PR is welcome – thanks
